### ISSUE TITLE
Inference is very slow on an TPU v4 instance - 2 Tokens / Sec 

### ISSUE THREAD

**Issue by rakesh010101** [erfanzar/EasyDeL#175]
**Describe the bug**

There seems to be an issue with partitioning and sharding of the model. When running llama-3.2-1b model, we are getting 2tokens/sec rate.

```
poetry run python 1.py 
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
E0000 00:00:1730813951.107825 3498419 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1730813951.112641 3498419 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
Converting Model: 100%|██████████████████████████████████████████████████████████████████████| 147/147 [00:49<00:00,  3.00it/s]
Inference model name: llama-1.50B-20241105
Running Inference
SampleState(
  current_length : 5221
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [3764348679  739019745]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 12721965261.56501
  tokens_pre_second : 3.4918914
  generated_tokens : 101
)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!SampleState(
  current_length : 5321
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [ 321572077 1027014635]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 59763692854.78287
  tokens_pre_second : 2.9158776
  generated_tokens : 201
)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!SampleState(
  current_length : 5421
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [ 498670314 1394438950]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 190537339899.5581
  tokens_pre_second : 2.7686794
  generated_tokens : 301
)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!SampleState(
  current_length : 5521
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [3279999784 3066355674]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 252607197372.8961
  tokens_pre_second : 2.6999717
  generated_tokens : 401
)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!SampleState(
  current_length : 5621
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [ 719919651 4067907500]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 288690996592.2975
  tokens_pre_second : 2.6605105
  generated_tokens : 501
)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!SampleState(
  current_length : 5721
  sequences : [[128009 128009 128009 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [2207948256 3561960881]
  model_kwargs : dict(...)
  generate_func_flops : 107181058509720.72
  interval_func_flops : 315473819222.4213
  tokens_pre_second : 2.6346762
  generated_tokens : 601
)

```

**To Reproduce**
Run the following code on TPU V4 instance

```
# Import necessary modules
import easydel as ed
from transformers import AutoTokenizer
import jax
from jax import numpy as jnp, sharding, lax

# Model configuration
model_name = "meta-llama/Llama-3.2-1B-Instruct"
max_length = 512
dtype = jnp.float16
sharding_axis_dims = (1, 1, 1, -1)
num_devices = len(jax.devices())
input_shape = (1, max_length)
partition_axis = ed.PartitionAxis()

# Load and configure model with EasyDeL
model, params = ed.AutoEasyDeLModelForCausalLM.from_pretrained(
    model_name,
    input_shape=input_shape,
    auto_shard_params=True,
    sharding_axis_dims=sharding_axis_dims,
    config_kwargs={
        "use_scan_mlp": False,
        "partition_axis": partition_axis,
        "attn_dtype": dtype,
        "freq_max_position_embeddings": max_length,
        "mask_max_position_embeddings": max_length,
        "attn_mechanism": ed.AttentionMechanisms.FLASH_ATTN2,
    },
    #quantization_method="8bit",
    platform="jax",
    param_dtype=dtype,
    dtype=dtype,
    precision=lax.Precision("fastest"),
)

# Initialize the tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name)
tokenizer.padding_side = "left"
tokenizer.pad_token_id = tokenizer.eos_token_id

# Define asynchronous inference function
async def run_inference():
    # Configure inference settings
    inference = ed.vInference(
        model=model,
        params=params,
        tokenizer=tokenizer,
        generation_config=ed.vInferenceConfig(
            max_new_tokens=1024,
            temperature=model.generation_config.temperature,
            top_p=model.generation_config.top_p,
            top_k=model.generation_config.top_k,
            eos_token_id=model.generation_config.eos_token_id,
            streaming_chunks=100,
        ),
    )

    # Precompile model for asynchronous inference
    await inference.async_precompile(1)
    print(f"Inference model name: {inference.inference_name}")

    # Prepare input for generation
    chat_input = [{"role": "user", "content": "Generate a long random story"}]
    ids = tokenizer.apply_chat_template(
        chat_input,
        return_tensors="np",
        return_dict=True,
        max_length=inference.model_prefill_length,
        padding="max_length",
        add_generation_prompt=True,
    )

    input_ids, attention_mask = ids["input_ids"], ids["attention_mask"]
    pad_seq = inference.model_prefill_length
    print("Running Inference")
    # Run asynchronous inference and output results
    async for response in inference.generate(
        input_ids=input_ids,
        attention_mask=attention_mask,
    ):
        print(response)
        next_slice = slice(
            pad_seq,
            pad_seq + inference.generation_config.streaming_chunks,
        )
        pad_seq += inference.generation_config.streaming_chunks
        print(
            tokenizer.decode(response.sequences[0][next_slice], skip_special_tokens=True),
            end="",
        )

# To execute the async function, you would need an async environment, such as:
import asyncio
asyncio.run(run_inference())

```

**Comment by erfanzar** [comment]
Hello
The issue your facing at the moment is caused by using sharding sequences and using flash attention 2 
You can not use both flash attention 2 and sequence sharding 
Please check that with vanilla attention 

Let me know if the issue still remains and ill look deeper

**Comment by rakesh010101** [comment]
I changed the code and now tokens / sec has increased, however, it has started giving junk results. 

```
Converting Model: 100%|█████████████████████████████████████████████████████████████████| 147/147 [00:52<00:00,  2.79it/s]
Inference model name: llama-1.50B-20241105
platform='tpu' 0 : available=30.060628992 | bytes_in_use=0.982832128
platform='tpu' 1 : available=31.144619008000003 | bytes_in_use=0.974414336
platform='tpu' 2 : available=31.144619008000003 | bytes_in_use=0.974414336
platform='tpu' 3 : available=31.144619008000003 | bytes_in_use=0.974414336
Running Inference
SampleState(
  current_length : 1125
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[31]]
  is_sequence_finished : [False]
  prng_key : [3764348679  739019745]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 16249291181.337942
  tokens_pre_second : 929.37164
  generated_tokens : 101
)
.)1/4!'1N;KM#7FD'%(B88=<44BK?7;?4LI36J-$=@$=,;&ACDAI@C1+B*('C";%6#%695H*94&*L%I%4N$97,/'/G2+B>F6<CASampleState(
  current_length : 1225
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[17]]
  is_sequence_finished : [False]
  prng_key : [ 321572077 1027014635]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 66612622961.855415
  tokens_pre_second : 141.0874
  generated_tokens : 201
)
@BNA67'#:G1F5K9"&>#(060>2G>3+++:7M01F;"(?1+F)'E!KIF2EA+L:%F/?=BA)&A078,L@8K1"9&)3N*FF296F@3=54,,0:3LSampleState(
  current_length : 1325
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[2]]
  is_sequence_finished : [False]
  prng_key : [ 498670314 1394438950]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 180986292688.065
  tokens_pre_second : 119.704605
  generated_tokens : 301
)
2(%C;,=?M1MFM1<'"F08H*-31;JN<#56!C"7F0*%F+.:3E+G-*,G:<+.:85;H,D36"KB+H>8)@/#8'0$+=JIK;=L5*)I#"+8.)C3SampleState(
  current_length : 1425
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[44]]
  is_sequence_finished : [False]
  prng_key : [3279999784 3066355674]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 234515100345.958
  tokens_pre_second : 111.15918
  generated_tokens : 401
)
#?L;)M'N?!):4I?@-8LG:!6''5-ME*B#KA4/D4$1>9GK2(763+)<CM%$/<NIGKH&K&H:1@*=3+2=31)&N(ADIN:-$B#C"9>/,<#(SampleState(
  current_length : 1525
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[35]]
  is_sequence_finished : [False]
  prng_key : [ 719919651 4067907500]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 268427671911.0499
  tokens_pre_second : 106.80852
  generated_tokens : 501
)
MD##4M8K@B0?LGCL+9%@A*/9#N,L4C&;+95!(66<(H6./@1,!83B0@8J=9K2G+H6!H>9G<4=#CEC0>0&BHNIJ-L2*(>731ML2K2)SampleState(
  current_length : 1625
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[19]]
  is_sequence_finished : [False]
  prng_key : [2207948256 3561960881]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 289693055296.66095
  tokens_pre_second : 104.10299
  generated_tokens : 601
)
D4&8!E75-8&12D=2EG'M>><.76D7%,'K;D&!N3I<DF6&J1.H76H2,I8"MB<43FDH)<FM@FH1M6E23N?H25C@#!./IH&:#A;1M;'NSampleState(
  current_length : 1725
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[32]]
  is_sequence_finished : [False]
  prng_key : [1485054744  454425506]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 306272426873.12866
  tokens_pre_second : 102.270645
  generated_tokens : 701
)
4!&>$CAKA&M:888--,/#>?L3@3KF;A!2FL,:4+B+918>?=FA>#?;%.EH79M(/#?H$4.I/F!E.>EI8!.C1.L<?4M$EL+B1!<60%,<SampleState(
  current_length : 1825
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[0]]
  is_sequence_finished : [False]
  prng_key : [2628581531  453362355]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 315218709679.79395
  tokens_pre_second : 100.90708
  generated_tokens : 801
)
A=;FF'!5'*7LE;4H32<1*%52#<62..**1D:#/!B*'FJ4/M'-&I9:I..0?/H(>H;C0)L54&$D.4<AI+C@EH-CL"E-6?KHJ=$10N!7SampleState(
  current_length : 1925
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[6]]
  is_sequence_finished : [False]
  prng_key : [2080234625 2554801132]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 323269828777.36084
  tokens_pre_second : 99.851036
  generated_tokens : 901
)
!.134:!G6!<1GKJ7?<A-<%0<*-2:.D/20<!*NL(>1N2A7!<J;ED$<BG+FK*N5)/!:F2?:N!/!+J2#E5&E?(6*"+N7N@!25(A&7J4SampleState(
  current_length : 2025
  sequences : [[128000 128006   9125 ... 128009 128009 128009]]
  running_token : [[44]]
  is_sequence_finished : [False]
  prng_key : [2469344182 3207131215]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 330807221202.23364
  tokens_pre_second : 99.079994
  generated_tokens : 1001
)
'DD<,K4+=<?I29)JM%9/1&6E?L)=I4/DI6!A2=(<3#:G=2E%LFL(-%>&$3%47-DK8069,EE%(H90*('B+95;%"A%H1++30I2)@"@SampleState(
  current_length : 2125
  sequences : [[128000 128006   9125 ...     23     44     42]]
  running_token : [[9]]
  is_sequence_finished : [False]
  prng_key : [3783360967 4099849946]
  model_kwargs : dict(...)
  generate_func_flops : 26290839454235.395
  interval_func_flops : 337312141668.11975
  tokens_pre_second : 98.41567
  generated_tokens : 1101
)

```


Here is the code 

```
# Import necessary modules
import easydel as ed
from transformers import AutoTokenizer
import jax
from jax import numpy as jnp, sharding, lax

# Model configuration
model_name = "meta-llama/Llama-3.2-1B-Instruct"
max_length = 2048
dtype = jnp.float16
sharding_axis_dims = (1, 1, 1, -1)
num_devices = len(jax.devices())
input_shape = (1, max_length)
partition_axis = ed.PartitionAxis()

# Load and configure model with EasyDeL
model, params = ed.AutoEasyDeLModelForCausalLM.from_pretrained(
    model_name,
    input_shape=input_shape,
    auto_shard_params=True,
    sharding_axis_dims=sharding_axis_dims,
    sharding_axis_names=("dp", "fsdp", "tp", "sp"),
    config_kwargs={
        "use_scan_mlp": False,
     #   "partition_axis": partition_axis,
        "attn_dtype": dtype,
        "freq_max_position_embeddings": max_length,
        "mask_max_position_embeddings": max_length,
        "attn_mechanism": ed.AttentionMechanisms.VANILLA,
    },
    #quantization_method="8bit",
    platform="jax",
    param_dtype=dtype,
    dtype=dtype,
    from_torch=True,
    precision=lax.Precision("fastest"),
)

# Initialize the tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name)
#tokenizer.padding_side = "right"
tokenizer.pad_token_id = tokenizer.eos_token_id


def accelerator_status(): # utiliti function.
    for idx, device in enumerate(jax.devices()):
        platform = device.platform
        status = device.memory_stats()
        bytes_in_use = status["bytes_in_use"] / 1e9
        available  = status["bytes_reservable_limit"] / 1e9 - bytes_in_use
        print(f"{platform=} {idx} : {available=} | {bytes_in_use=}")

# Define asynchronous inference function
async def run_inference():
    # Configure inference settings
    inference = ed.vInference(
        model=model,
        params=params,
        tokenizer=tokenizer,
        generation_config=ed.vInferenceConfig(
            max_new_tokens=1024,
            temperature=model.generation_config.temperature,
            top_p=model.generation_config.top_p,
            top_k=model.generation_config.top_k,
            eos_token_id=model.generation_config.eos_token_id,
            streaming_chunks=100,
        ),
    )

    # Precompile model for asynchronous inference
    await inference.async_precompile(1)
    print(f"Inference model name: {inference.inference_name}")

    # Prepare input for generation
    chat_input = [{"role": "user", "content": "Hello How are you?"}]
    ids = tokenizer.apply_chat_template(
        chat_input,
        return_tensors="np",
        return_dict=True,
        max_length=inference.model_prefill_length,
        padding="max_length",
        add_generation_prompt=True,
    )

    input_ids, attention_mask = ids["input_ids"], ids["attention_mask"]
    pad_seq = inference.model_prefill_length
    accelerator_status()
    print("Running Inference")
    # Run asynchronous inference and output results
    async for response in inference.generate(
        input_ids=input_ids,
        attention_mask=attention_mask,
    ):
        print(response)
        next_slice = slice(
            pad_seq,
            pad_seq + inference.generation_config.streaming_chunks,
        )
        pad_seq += inference.generation_config.streaming_chunks
        print(
            tokenizer.decode(response.sequences[0][next_slice], skip_special_tokens=True),
            end="",
        )

# To execute the async function, you would need an async environment, such as:
import asyncio
asyncio.run(run_inference())


```

Can you please suggest how to use your library to make faster inferences from Llama models on TPUs? please suggest few examples.

**Comment by erfanzar** [comment]
Yea ill check that asap

**Comment by rakesh010101** [comment]
Sure, it will be helpful.

**Comment by erfanzar** [comment]
sorry for being late to answer your question
actually everything is fine in your code except attn_dtype which should be in float32 since tpus dont support float16 and bfloat16 accurate computation
change that and eveything will start working just fine

**Comment by erfanzar** [comment]
and also set tokenizer padding side to left

**Comment by erfanzar** [comment]
@rakesh010101 is there any other issue?
